chapter3/calc/Eval.hs

```diff
@@ -34,6 +34,6 @@ nf :: Expr -> Expr
 nf x = fromMaybe x (nf <$> eval' x)
 
 eval :: Expr -> Maybe Expr
-eval t = case isVal (nf t) of
-  True  -> Just (nf t)
-  False -> Nothing -- term is "stuck"
+eval t = case nf t of
+  nft | isVal nft -> Just nft
+      | otherwise -> Nothing -- term is "stuck"
```

Comment:
That again recomputes the normal form.
I'd suggest:
```haskell
eval t =
  let nft = nf t in
    if isVal nft
    then nft
    else Nothing -- term is "stuck"
```
Before I tried to keep the style and line count.
(Forgot to click on "Comment" earlier...)